Yoast/wordpress-seo
# Saves the reason for the notice to show

## Context


* When the permalinks are reset for the indexables we want to show a notification. By saving the reason for reindexation we can provide a proper notice without having difficult logic.

## Summary


This PR can be summarized in the following changelog entry:

* Shows a more specific notification when the permalinks are reset or when the category base setting is changed

## Relevant technical choices:

* Replaced the test with the dataprovider, because it became too complex to cover both situations. Instead of a dataprovider with two scenarios I have made two tests that covers both cases.
* I decided to extend `Indexable_Category_Permalink_Watcher` with `Indexable_Permalink_Watcher` because I made equal changes to both classes for the third time and I think this is more practical and they represents an equal concept: resetting permalinks.

## Test instructions

This PR can be tested by following these steps:

* Use this piece of code to bring the notice back, it unsets some hide settings:
```php
add_action( 'init', function() { 
    WPSEO_Options::set( 'ignore_indexation_warning', false );
    WPSEO_Options::set( 'indexation_warning_hide_until', false );
    WPSEO_Options::set( 'indexation_started', false );
} );
```

You also need this to your theme functions.php file or some place else:
```php
add_filter( 'wpseo_shutdown_indexation_limit', function() { return 0; });
```

* Make sure you have categories and posts with permalinks
* Go to Yoast -> Search Appearance -> Taxonomies and toggle the setting _Remove the categories prefix_. You might need to scroll a little bit, it is at the bottom.
* Visit your tools page and make verify you have a notice that contains:
> Because of a change in your category URL setting, some of your SEO data need to be reprocessed. 
* Reindex the indexables
* Navigate to the permalink settings and make a change in it. This can be either the general permalink settings or category_base / tag_base.
* Visit the tools page again and see you have a notice that contains: 
> Because of a change in your permalink structure, some of your SEO data need to be reprocessed.
* Hide the notice and make sure it isn't visible anymore.
* Reindex the indexables.
* Change the permalinks or category base setting again, the notice should reappear.

## UI changes

## Documentation

## Quality assurance


Fixes #15440
